docs/core/v3.0.rst

```diff
@@ -0,0 +1,1762 @@
+resulting compression ratio of the data. Configuration parameters must
+be declared in the codec specification, including a definition of how
+configuration parameters are represented as JSON.
+
+The Zarr core development team maintains a repository of codec
+specifications, which are hosted alongside this specification in the
+`zarr-specs GitHub repository`_, and which are
+published on the `zarr-specs documentation Web site
+<http://zarr-specs.readthedocs.io/>`_. For ease of discovery, it is
+recommended that codec specifications are contributed to the
+zarr-specs GitHub repository. However, codec specifications may be
+maintained by any group or organisation and published in any location
+on the Web. For further details of the process for contributing a
+codec specification to the zarr-specs GitHub repository, see the Zarr
+community process specification.
+
+Further details of how a compressor is configured for an array are
+given in the section below on `Array metadata`_.
+
+
+Metadata
+========
+
+This section defines the structure of metadata documents for Zarr hierarchies,
+which consists of three types of metadata documents: an entry point metadata
+document (``zarr.json``), array metadata documents, and group metadata
+documents. Each type of metadata document is described in the following
+subsections.
+
+Metadata documents are defined here using the JSON
+type system defined in [RFC8259]_. In this section, the terms "value",
+"number", "string" and "object" are used to denote the types as
+defined in [RFC8259]_. The term "array" is also used as defined in
+[RFC8259]_, except where qualified as "Zarr array". Following
+[RFC8259]_, this section also describes an object as a set of
+name/value pairs. This section also defines how metadata documents are
+encoded for storage.
+
+
+Only the top level metadata document ``zarr.json`` is guaranteed to be of JSON
+type, and can be used to define other formats for array-level and group-level
+metadata documents. In the case where non-JSON metadata documents are used in a
+Zarr hierarchy, the following sections on group and array level metadata are
+non-normative, but other metadata formats are expected to define some
+equivalence relations with the JSON documents.
+
+
+Entry point metadata
+--------------------
+
+Each Zarr hierarchy must have an entry point metadata document, which
+provides essential information regarding the format version being
+used, the encoding being used for group and array metadata, and any
+extensions that affect the layout or interpretation of data
+in the store.
+
+The entry point metadata document must contain a single object
+containing the following names:
+
+``zarr_format``
+^^^^^^^^^^^^^^^
+
+    A string containing the URI of the Zarr core
+    specification that defines the metadata format. For Zarr
+    hierarchies conforming to this specification, the value must be
+    the string "https://purl.org/zarr/spec/core/3.0".
```

Comment:
Currently zarr v3 uses a mixture of full URLs and short strings as identifiers.  I can see that there are some advantages to using full URLs: rely on existing DNS registration mechanism so there is no need for a separate "name reservation" process, automatic linking to a corresponding explanatory document assuming that the URLs actually remain valid.

However, I think the long URLs would be cumbersome for a user to actually type manually, and even just looking at the metadata, it will be harder to read at a glance, since e.g. all of the codecs will start with the same long prefix, and just the part at the end is relevant.  There are also surely a lot of cases where the user will want to specify something that is currently identified only by a URL, e.g. the codec to use, the metadata format to use, the grid to use, the storage transformer to use, etc.

In order to save users from having to type out full URLs to identify these things, implementations may be tempted to invent their own short identifiers as well, and it may be confusing if these identifiers conflict, and there will be the added complexity of having to relate the short identifiers to the full URLs.

Reply:
It depends on the implementation, but my guess is that users often won't type these full URLs directly (moreover the implementation will likely fill in this metadata).

Reply:
Yeah, it depends on the implementation.  In TensorStore, for example, there is a general format-independent "schema" representation for specifying certain properties, but users can also just directly specify the zarr metadata in order to precisely control properties of the array:

```json
{"driver": "zarr",
  "kvstore": "file://...",
  "metadata": {"order": "F", "dtype": ">u2", ...}
}
```

TensorStore does not bother to define its own alternative representation of the zarr v2 metadata, since the zarr v2 metadata is already reasonably concise.  It would be nice not to have to define an alternative more concise representation of the zarr v3 metadata, as that would likely lead to added confusion.

More generally, users will need some way to specify everything that is in the metadata when creating a zarr repo/array, ideally in a concise way.  If the zarr core spec or extension specs do not provide a concise representation, then each implementation may need to invent its own concise representation, which would be unfortunate.

Reply:
I extracted this to issue #176, please feel free to comment there.